### PR TITLE
ansible: check ansible exists or not before test start

### PR DESCRIPTION
**Pull request by yanan-fu**
1. Check ansible exists or not before test start
2. Install ansible use the method that defined by ansible_install_policy
3. If install ansible with pip, install sshpass separately as it can't be
    installed automatically. Install specified sshpass if failed to install
    it from distro and there is a specified sshpass pkg.

id: 1993730

Signed-off-by: Yanan Fu <yfu@redhat.com>

### CODE REVIEW

**Comment by yanan-fu** [comment]
@PaulYuuu  I changed the design with the latest version.
1. old:  checking ansible when import the module, and then sshpass in ansible_test as it need `params`
2. new: add new function in `provider.ansible` to check both ansible and sshpass,  call it in `ansible_test` as the first step, cancel test if check failed.

Could you help review ? Thanks!

**Comment by yanan-fu** [comment]
Test log for key steps:
```
04:36:00 INFO | Running 'rpm -q ansible'
04:36:00 DEBUG| [stdout] package ansible is not installed
04:36:00 INFO | Command 'rpm -q ansible' finished with 1 after 0.012918494998302776s
04:36:00 INFO | Running '/usr/bin/dnf -y install ansible'
...
04:36:01 DEBUG| [stdout] No match for argument: ansible
04:36:01 DEBUG| [stderr] Error: Unable to find a match: ansible
04:36:01 INFO | Command '/usr/bin/dnf -y install ansible' finished with 1 after 0.5848911149951164s
04:36:01 ERROR| Operate ansible on host failed
04:36:02 INFO | Running 'which pip3'
04:36:02 DEBUG| [stdout] /usr/local/bin/pip3
04:36:02 INFO | Command 'which pip3' finished with 0 after 0.001815177995013073s
04:36:02 INFO | Running 'pip3 install ansible'
...
04:37:01 INFO | Command 'pip3 install ansible' finished with 0 after 59.3651554790049s
04:37:01 INFO | Running 'rpm -q sshpass'
04:37:01 DEBUG| [stdout] package sshpass is not installed
04:37:01 INFO | Command 'rpm -q sshpass' finished with 1 after 0.007484848996682558s
04:37:01 INFO | Running '/usr/bin/dnf -y install sshpass'
...
04:37:03 INFO | Command '/usr/bin/dnf -y install sshpass' finished with 0 after 1.1953109160021995s
...
04:37:05 INFO | Running qemu command (reformatted):
```

**Comment by yanan-fu** [comment]
@qcheng-redhat Could you help review ? Thanks!

File: provider/ansible.py [hunk 1]
```diff
@@ -116,3 +107,45 @@ def store_playbook_log(self, log_dir, filename):
...
+        :param try_binaries: A list of binaries names to look for.
+        :return: The binary found, otherwise None.
+        """
+        for binary in try_binaries:
+            if process.system("which %s" % binary, ignore_status=True) == 0:
+                return binary
+
+    try:
+        path.find_command('ansible-playbook')
+    except path.CmdNotFoundError:
```
Comment (PaulYuuu) [inline provider/ansible.py:133]: Maybe we can let users choose how to install the ansible.

**Comment by yanan-fu** [comment]
@PaulYuuu @qcheng-redhat  Update the policy as we discussed offline based on the new product change.
Could you help review, thanks!

File: provider/ansible.py [hunk 2]
```diff
@@ -116,3 +107,54 @@ def store_playbook_log(self, log_dir, filename):
...
+        """
+        if not utils_package.package_install('ansible'):
+            return False
+        return True
+
+    policy_map = {"distro_install": distro_install,
+                  "python_install": python_install}
+
+    try:
+        path.find_command('ansible-playbook')
```
Comment (qcheng-redhat) [inline provider/ansible.py:157]: If ansible-playbook found, needs return.

File: provider/ansible.py [hunk 3]
```diff
@@ -116,3 +107,54 @@ def store_playbook_log(self, log_dir, filename):
...
             log_file.write(self.get_output())
             log_file.flush()
+
+
+def check_ansible_playbook(params):
+    """
+    check if ansible playbook exist or not.
+
+    :param params: Dictionary with the test parameters.
+    :return: True if ansible playbook exist or be installed success, else False.
```
Comment (qcheng-redhat) [inline provider/ansible.py:117]: ansible playbook -> ansible-playbook

File: provider/ansible.py [hunk 4]
```diff
@@ -116,3 +107,54 @@ def store_playbook_log(self, log_dir, filename):
         with open(os.path.join(log_dir, filename), 'w') as log_file:
             log_file.write(self.get_output())
             log_file.flush()
+
+
+def check_ansible_playbook(params):
+    """
+    check if ansible playbook exist or not.
```
Comment (qcheng-redhat) [inline provider/ansible.py:114]: ansible playbook -> ansible-playbook

File: provider/ansible.py (hunk 2 above)
Comment (yanan-fu) [reply provider/ansible.py:157]: Good catch, updated, thanks~

File: provider/ansible.py [hunk 5]
```diff
@@ -116,3 +107,56 @@ def store_playbook_log(self, log_dir, filename):
...
+    """
+
+    def python_install():
+        """
+        Install python ansible.
+        """
+        pip_bin = ''
+        for binary in ['pip', 'pip3', 'pip2']:
+            if process.system("which %s" % binary, ignore_status=True) == 0:
+                pip_bin = binary
```
Comment (PaulYuuu) [inline provider/ansible.py:127]: Need to break the loop when the pip is found.

**Comment by qcheng-redhat** [comment]
LGTM.

On rhel 8:

02:22:51 INFO | Running '/bin/dnf -y install ansible'
02:22:51 DEBUG| [stdout] Updating Subscription Management repositories.
......
 ================================================================================
02:22:52 DEBUG| [stdout]  Package           Architecture     Version                Repository      Size
02:22:52 DEBUG| [stdout] ================================================================================
02:22:52 DEBUG| [stdout] Installing:
02:22:52 DEBUG| [stdout]  ansible           noarch           2.9.24-1.el8           epel            17 M
02:22:52 DEBUG| [stdout] Installing dependencies:
02:22:52 DEBUG| [stdout]  sshpass           x86_64           1.06-9.el8             epel            27 k
02:22:52 DEBUG| [stdout]
02:22:52 DEBUG| [stdout] Transaction Summary
02:22:52 DEBUG| [stdout] ================================================================================
02:22:52 DEBUG| [stdout] Install  2 Packages
02:22:52 DEBUG| [stdout]
02:22:52 DEBUG| [stdout] Total download size: 17 M
02:22:52 DEBUG| [stdout] Installed size: 96 M
02:22:52 DEBUG| [stdout] Downloading Packages:
02:22:53 DEBUG| [stdout] (1/2): sshpass-1.06-9.el8.x86_64.rpm            307 kB/s |  27 kB     00:00
02:22:56 DEBUG| [stdout] (2/2): ansible-2.9.24-1.el8.noarch.rpm          4.5 MB/s |  17 MB     00:03
02:22:56 DEBUG| [stdout] --------------------------------------------------------------------------------
02:22:56 DEBUG| [stdout] Total                                           3.8 MB/s |  17 MB     00:04
......
02:23:01 DEBUG| [stdout]
02:23:01 DEBUG| [stdout] Installed:
02:23:01 DEBUG| [stdout]   ansible-2.9.24-1.el8.noarch             sshpass-1.06-9.el8.x86_64
02:23:01 DEBUG| [stdout]
02:23:01 DEBUG| [stdout] Complete!
02:23:01 INFO | Command '/bin/dnf -y install ansible' finished with 0 after 10.199762610951439s

On rhel 9:
......
01:59:28 INFO | Running 'pip3 install ansible'
01:59:28 DEBUG| [output] Requirement already satisfied: ansible in /usr/local/lib/python3.9/site-packages (4.4.0)
......
01:59:34 DEBUG| [output] Successfully installed ansible-core-2.11.4
01:59:34 DEBUG| Successfully installed ansible-core-2.11.4
01:59:34 INFO | Command 'pip3 install ansible' finished with 0 after 6.666453861980699s
01:59:34 INFO | Running 'rpm -q sshpass'
01:59:34 DEBUG| [stdout] package sshpass is not installed
01:59:34 INFO | Command 'rpm -q sshpass' finished with 1 after 0.006649928021943197s
01:59:34 INFO | Running '/usr/bin/dnf -y install sshpass'
......
01:59:36 DEBUG| [stdout] Installed:
01:59:36 DEBUG| [stdout]   sshpass-1.09-2.el9.x86_64
01:59:36 DEBUG| [stdout]
01:59:36 DEBUG| [stdout] Complete!
01:59:36 INFO | Command '/usr/bin/dnf -y install sshpass' finished with 0 after 1.4225569129921496s

File: provider/ansible.py (hunk 5 above)
Comment (yanan-fu) [reply provider/ansible.py:127]: Good catch, updated, thanks!

**Comment by yanan-fu** [comment]
@PaulYuuu Do you have any more comments ? Thanks!

**Review by PaulYuuu — APPROVED** [review]

**Comment by yanan-fu** [comment]
@qcheng-redhat  @PaulYuuu  Any new comment with the latest update ? Thanks!

**Comment by qcheng-redhat** [comment]
No more comment from my side. Thanks.

**Comment by PaulYuuu** [comment]
Still LGTM, let's merge it accordingly.

**Pull request merged**